src/control/lib/hardware/fabric.go

```diff
@@ -859,23 +793,39 @@ func (s *FabricScanner) CacheTopology(t *Topology) error {
 	return nil
 }
 
-// FabricReadyChecker is an interface for a type that can be used to check whether a fabric
-// interface is ready.
-type FabricReadyChecker interface {
-	CheckFabricReady(string) error
+// NetDevState describes the state of a network device.
+type NetDevState int
+
+const (
+	// NetDevStateUnknown indicates that the device state can't be determined.
+	NetDevStateUnknown NetDevState = iota
+	// NetDevStateDown indicates that the device is down. This could mean the device is disabled
+	// or physically disconnected.
+	NetDevStateDown
+	// NetDevStateNotReady indicates that the device is connected but not ready to communicate.
+	NetDevStateNotReady
+	// NetDevStateReady indicates that the device is up and ready to communicate.
+	NetDevStateReady
+)
+
+// NetDevStateProvider is an interface for a type that can be used to get the state of a network
+// device.
+type NetDevStateProvider interface {
+	GetNetDevState(string) (NetDevState, error)
 }
 
 // WaitFabricReadyParams defines the parameters for a WaitFabricReady call.
 type WaitFabricReadyParams struct {
-	Checker        FabricReadyChecker
+	StateProvider  NetDevStateProvider
 	FabricIfaces   []string      // Fabric interfaces that must become ready
 	IterationSleep time.Duration // Time between iterations if some interfaces aren't ready
+	IgnoreUnusable bool          // Ignore interfaces that are in an unusable state
```

Comment:
Hmm. This is likely to be the most common case, right? If so, it might be better to change this so that the zero value of the struct defaults to the most common behavior rather than making every caller have to set this to true. e.g. `IgnoreUnusable` -> `IncludeUnusable` so that a caller that really wants to include unusable interfaces in the criteria can do so.

Speaking of which, when/why would a caller want to do that? If there's no sensible use case for it, maybe it shouldn't even be exposed as a parameter?

Reply:
Maybe I could clarify the naming. Essentially this one means "don't wait or error on listed interfaces that are down." It is specifically for the agent use case.

The server startup flow is basically:
- Get list of engine interfaces from config file.
- Wait for all of those interfaces to come up before proceeding. We cannot ignore interfaces that are unplugged or disabled. The server needs them.

The agent needs to make sure the fabric is up when dealing with the first client request:
- GetAttachInfo from server. This contains the network device class.
- Get at all local client net interfaces that satisfy the device class. These may be in any state.
- We have no idea which of these we'll use yet, but there is no point in scanning the fabric until we're sure they're up and will have their domains detected by libfabric/ucx. The ones in the Down state are either disabled or unplugged, intentionally or unintentionally. They may never come up. So we ignore them and wait for the others to come up before proceeding.

I know there are some customer configs that explicitly disable one port or one NIC.

Reply:
It depends a bit. The interface will still be detected on fabric scan, but if we're using verbs (or anything where we need a domain), we won't have the domain. If we're using tcp, I would expect it to just work.